src/metabase/api/query_metadata.clj

```diff
@@ -1,162 +1,147 @@
+                  (update m :cards conj! card-id)
+                  (update m :tables conj! src)))
+              {:cards  (transient #{})
+               :tables (transient #{})}
+              source-ids)
+      (update-vals persistent!)))
 
-(defn- metadata-for-field-dependents
-  [field-dependents]
-  (let [metadatas (api.field/get-fields (map :id field-dependents))]
-    (when (< (count metadatas) (count field-dependents))
-      (log/warn "Some fields are not readable by the current user"))
-    (vec (sort-by :id metadatas))))
+(defn- query->template-tag-field-ids [query]
+  (when-let [template-tags (some-> query :native :template-tags vals seq)]
+    (for [{tag-type :type, [dim-tag id _opts] :dimension} template-tags
+          :when (and (= tag-type :dimension)
+                     (= dim-tag :field)
+                     (integer? id))]
+      id)))
 
-(defn- metadata-for-dependents [dependents]
-  {:tables (->> (:table dependents)
-                ;; Can be int or "card__<id>"
-                metadata-for-table-dependents)
-   :databases (->> (:database dependents)
-                   (sort-by :id)
-                   (into []
-                         (keep #(fetch-or-warn % api.database/get-database {}))))
-   :fields (->> (:field dependents)
-                metadata-for-field-dependents)})
+(defn- batch-fetch-query-metadata*
+  "Fetch dependent metadata for ad-hoc queries."
+  [queries]
+  (let [source-ids                (into #{} (mapcat #(lib.util/collect-source-tables (:query %)))
+                                        queries)
+        {source-table-ids :tables
+         source-card-ids  :cards} (split-tables-and-legacy-card-refs source-ids)
+        source-tables             (concat (api.table/batch-fetch-table-query-metadatas source-table-ids)
+                                          (api.table/batch-fetch-card-query-metadatas source-card-ids))
+        fk-target-field-ids       (into #{} (comp (mapcat :fields)
+                                                  (keep :fk_target_field_id))
+                                        source-tables)
+        fk-target-table-ids       (into #{} (remove source-table-ids)
+                                        (field-ids->table-ids fk-target-field-ids))
+        fk-target-tables          (api.table/batch-fetch-table-query-metadatas fk-target-table-ids)
+        tables                    (concat source-tables fk-target-tables)
+        template-tag-field-ids    (into #{} (mapcat query->template-tag-field-ids) queries)
+        query-database-ids        (into #{} (keep :database) queries)
+        database-ids              (into query-database-ids
+                                        (keep :db_id)
+                                        tables)]
+    {;; TODO: This is naive and issues multiple queries currently. That's probably okay for most dashboards,
+    ;; since they tend to query only a handful of databases at most.
```

Comment:
This comment actually doesn't apply anymore, but it's not worth another CI run to remove it.